### PR TITLE
KFServer & XGBoost Server

### PR DESCRIPTION
**Pull request by ellis-bigelow**
# TL;DR
Very WIP. Please comment ideas if you think the high level idea is flawed. Please let me know if you have any ideas for structuring the packages re: dependencies. 

# Main Idea
1. Flexible Library of Python files for use within KFServing or extension by Users
2. XGBoost, SKLearn, Pytorch, and other OOTB Python servers should implement KFServer
3. KFServer should implement utilities for retrieving pickle/saved_model files from all cloud providers
4. KFServer should make it easy to implement custom Predict functions extended by users
5. Initial Implementation will use Tensorflow Serving's API and support HTTP only.

# Remaining Work
1. I want to write a doc discussing the high level strategy here regarding protocols, will probably co-author with Clive
2. Get XGBoost working w/ GCS
3. Clean up code comments, file structure, lots of other things

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/61)

### CODE REVIEW

**Comment by rakelkar** [comment]
hmm it appears I am not a kubeflow member...?

File: python/kfserving/kfserver/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from .kfserver import KFServer
+from .kfmodel import KFModel
+from .storage import Storage
```
Comment (animeshsingh) [inline python/kfserving/kfserver/__init__.py]: nitpick - extra space

File: python/setup.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+setup(
+    name='kfserving',
+    version='0.1.0',
+    author_email='ellisbigelow@google.com',
+    license='../../LICENSE.txt',
+    url='https://github.com/kubeflow/kfserving/model-servers/kfserver',
+    description='Model Server for arbitrary python ML frameworks.',
+    long_description=open('README.md').read(),
+    python_requires='>3.4',
```
Comment (rakelkar) [inline python/setup.py]: add these files under a versioned path so that support for other python versions can also be implemented?

File: python/setup.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    packages=find_packages(),
+    install_requires=[
+        "tornado >= 1.4.1",
+        "xgboost == 0.82",
```
Comment (rakelkar) [inline python/setup.py]: base package should only include tornado since scikit images may not need xgboost (and possibly vice versa)?

File: python/kfserving/kfserver/kfserver.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    def __init__(self):
+        self.http_port = args.http_port
+        self.grpc_port = args.grpc_port
+        self.protocol = DEFAULT_PROTOCOL
+
+        self.model_name = args.model_name
+        self.model_dir = args.model_dir
+        self.local_model_dir = args.local_model_dir
+
+    def start(self, models):
```
Comment (animeshsingh) [inline python/kfserving/kfserver/kfserver.py]: Is it 1:1 mapping between server and model> self.model_name gives that impression, but then start takes multiple models...

File: python/kfserving/kfserver/storage.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class Storage(object):
+    @staticmethod
+    def download(uri, local_dir):
+        if uri.startswith(_GCS_PREFIX):
+            Storage._download_gcs(uri, local_dir)
+        elif uri.startswith(_S3_PREFIX):
+            Storage._download_s3(uri, local_dir)
+        elif uri.startswith(_LOCAL_PREFIX):
```
Comment (animeshsingh) [inline python/kfserving/kfserver/storage.py]: what does local here refer to? and are we accounting for storage like PV etc?

File: python/kfserving/kfserver/storage.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import shutil
+import errno
+import logging
+
+_GCS_PREFIX = "gs://"
+_S3_PREFIX = "s3://"
+_LOCAL_PREFIX = "/"
+
+
+class Storage(object):
```
Comment (rakelkar) [inline python/kfserving/kfserver/storage.py]: wonder if its possible to inject an init container to download stuff.. will allow this to be independent of storage implementations and will also potentially enable caching... just a thought possibly for a later date?

File: python/kfserving/xgbserver/__main__.py [hunk 7]
```diff
@@ -0,0 +1,4 @@
+from server import XGBoostServer
+
+def __main__():
+    XGBoostServer()
```
Comment (animeshsingh) [inline python/kfserving/xgbserver/__main__.py]: nitpick - extra space

File: python/setup.py (hunk 2 above)
Comment (ellis-bigelow) [reply python/setup.py]: I was thinking we use git branching for this. WDYT?

I was also thinking each kfserver release version maps to a framework version tuple (f1v1, f2v2, f3v1).

File: python/setup.py (hunk 3 above)
Comment (ellis-bigelow) [reply python/setup.py]: I probably need to make separate setup.py's for different frameworks. Alternatively we bundle them all together, but it may introduce unnecessary constraints. I'm not really worried about image size.

File: python/kfserving/kfserver/storage.py (hunk 6 above)
Comment (ellis-bigelow) [reply python/kfserving/kfserver/storage.py]: Caching with Persistent Volumes?

File: python/kfserving/kfserver/kfserver.py (hunk 4 above)
Comment (ellis-bigelow) [reply python/kfserving/kfserver/kfserver.py]: KFServer follows TensorflowServing's API of the ability of host multiple models. KFServer/XGBoost implements the server as a singleton.

File: python/kfserving/kfserver/storage.py (hunk 5 above)
Comment (ellis-bigelow) [reply python/kfserving/kfserver/storage.py]: Mostly for developer workflow, or if you happen to mount a PV. The PV should be able to be read directly using this.

File: python/kfserving/kfserver/kfserver.py (hunk 4 above)
Comment (yuzisun) [reply python/kfserving/kfserver/kfserver.py]: I think we are trying to support a model server that can host different versions of model right? The model map makes me a bit confused since that can contain different model names, but KFServer is initialized with a single model name.

File: python/setup.py (hunk 3 above)
Comment (ellis-bigelow) [reply python/setup.py]: Check the latest :)

File: python/kfserving/kfserver/kfserver.py (hunk 4 above)
Comment (ellis-bigelow) [reply python/kfserving/kfserver/kfserver.py]: I want the base server to support multi-models (like TFServing). The initial impl of XGBoostServer is just a singleton. I've cleaned this up in the code though.

File: python/kfserving/kfserver/storage.py (hunk 6 above)
Comment (rakelkar) [reply python/kfserving/kfserver/storage.py]: yes - we have some teams that have implemented this on k8s.. but the download should be decoupled from the caching obviously. In general will keep the container small and avoid having it implement download from all clouds (and keep up with versions)...

File: python/setup.py (hunk 2 above)
Comment (rakelkar) [reply python/setup.py]: I prefer paths as that makes it easier to see what is supported and make changes... is using git branches an established pattern used elsewhere? Also if we were to go down the git branches route then this should be in a repo of its own.. branches where only part of the code changes and other code is stale seems eeky

**Comment by animeshsingh** [comment]
@ellis-bigelow is this working? Good to get an initial working version merged, get the developer guide on, with some roadmap items indicated.

File: python/kfserving/server.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+        try:
+            body = json.loads(self.request.body)
+        except json.decoder.JSONDecodeError as e:
+            raise tornado.web.HTTPError(
+                status_code=HTTPStatus.BAD_REQUEST,
+                reason="Unrecognized request format: %s" % e
+            )
+
+        if "instances" not in body:
```
Comment (cliveseldon) [inline python/kfserving/server.py]: So we are creating a meta-schema that all protocols must adhere to? I thought we would be sending the protocol directly to the server?

**Comment by cliveseldon** [comment]
I think python version is quite a big issue. Ideally I would prefer we just support python >3.4 and have all the code use the latest features such as type annotations for best practice and clean code. I realise this means we could not serve python 2 sklearn python pickles unless someone knows a way to do this? Python 2 will not be maintained in 7 months time so I'm not sure handling old data science models in python 2 pickles just for sklearn is worth it - we should push people to upgrade in a similar way to saying we only support KNative if you want to use kfserving.

I realise this is maybe a contentious issue so lets discuss pros/cons.

File: python/kfserving/kfserver/kfserver.py (hunk 4 above)
Comment (cliveseldon) [reply python/kfserving/kfserver/kfserver.py]: How does handling different versions of models fit with the current control plane spec and "gitops"? 

I'm not sure we want to obscure what is being deployed. My feeling is TfServing came from a more monolithic serving architecture which expected to handle multiple models and versions and updates to them but what we are doing is more nimble where we handle canarys and updates at a higher level. I think it might be confusing if you have a server handling multiple models and you are also doing canary updates at a higher level. Am I missing something here or do you just mean a single model with multiple named inputs?

File: python/kfserving/kfserver/storage.py (hunk 6 above)
Comment (cliveseldon) [reply python/kfserving/kfserver/storage.py]: So for a side-car are we saying the servers always expect their models to be in a hardwired local location? As long as easy standalone testing is possible without the sidecar. Certainly all the issues with Authentication can then be handled by this separate piece of logic which might be good?

File: python/setup.py (hunk 2 above)
Comment (ellis-bigelow) [reply python/setup.py]: I would love to put this into a KFServing-core or KFServing-python repo.

Can we punt on this question for now? I want to enable an MVP. These sound like great feature requests and future work ;)

File: python/kfserving/kfserver/storage.py (hunk 6 above)
Comment (ellis-bigelow) [reply python/kfserving/kfserver/storage.py]: Here's an alternative thought. What if we used a knative build to bundle the model into the serving container?

Kubernetes (GKE especially) has lots of optimizations for image caching. We effectively get PVs for free. It also achieves decoupling the storage auth from the service.

File: python/kfserving/server.py (hunk 8 above)
Comment (ellis-bigelow) [reply python/kfserving/server.py]: as discussed in meeting, this is TF API

File: python/kfserving/kfserver/kfserver.py (hunk 4 above)
Comment (ellis-bigelow) [reply python/kfserving/kfserver/kfserver.py]: As discussed today, this is to enable model server multi-tenancy use cases. The KFServing interface will not need to change. We can enable this as an optimization in the future.

**Comment by ellis-bigelow** [comment]
> @ellis-bigelow is this working? Good to get an initial working version merged, get the developer guide on, with some roadmap items indicated.

I can run 
`python xgbserver` and then curl with the sample and it works. More effort needed to Dockerize it. I'll probably do it this weekend.

**Comment by ellis-bigelow** [comment]
Hey All. I think this is reasonably ready as a starting place.

File: python/kfserving/server.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
+from http import HTTPStatus
```
Comment (yuzisun) [inline python/kfserving/server.py]: missing license header here?

File: python/kfserving/storage.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
+import os
+import shutil
+import errno
```
Comment (yuzisun) [inline python/kfserving/storage.py]: these are not used

File: python/kfserving/storage.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import logging
+import random
```
Comment (yuzisun) [inline python/kfserving/storage.py]: random is not used

File: python/kfserving/server.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        self.model_name = args.model_name
+        self.model_dir = args.model_dir
+        self.local_model_dir = args.local_model_dir
+
+    def start(self, models=[]):
+        # TODO add a GRPC server
+        for model in models:
+            self.register_model(model)
+
+        self._http_server = tornado.web.Application([
```
Comment (yuzisun) [inline python/kfserving/server.py]: this is defined outside of` __init__`, perhaps in `__init__` we can add `self._http_server = None` ?

File: python/kfserving/storage.py (hunk 10 above)
Comment (ellis-bigelow) [reply python/kfserving/storage.py]: I need a better pylinter

File: python/kfserving/server.py (hunk 9 above)
Comment (ellis-bigelow) [reply python/kfserving/server.py]: Do you know how to generate these with the hack file?

File: python/kfserving/kfserver/storage.py (hunk 6 above)
Comment (rakelkar) [reply python/kfserving/kfserver/storage.py]: this thought seems at odds with the server multi-tenancy we discussed?

Comment (ellis-bigelow) [reply python/kfserving/kfserver/storage.py]: Agreed, but consider baking n models. It really depends on how dynamic we want the multi-tenancy loading/unloading to be. I worry that the eventual consistency of DNS-based routing will make it challenging for us to do too much dynamic loading/unloading. The problem with baking is that we may end up with combinatorial images (sparcely built) or huge images (densly built).

We can always support a lower cold start for single models, and the multi-model-server might suffer some tradeoffs.
